### PR TITLE
🌎 Set up common API types

### LINKED ISSUE #42: Static API types
# Purpose

It would be amazing to have shared, statically typed API types shared both by the client and server folders in our repo. Implementing this would be fully leveraging Typescript to its fullest.

# Resources
Not many resources on this. A lot of research would have to be done.

- Check out this [video](https://www.youtube.com/watch?v=GrnBXhsr0ng) by Gary Bernhardt on an end to end Typescript codebase (unfortunately it's not open source)
- https://lorefnon.tech/2018/03/25/typescript-and-validations-at-runtime-boundaries/
- io-ts might need to be used, or some sort of validator. Gary uses io-ts

### PR DESCRIPTION
**Pull request by xujustinj**
## Purpose

Closes #42 

We want a canonical representation of our API that both the frontend and backend can use. It's quite easy to write down a schema for our API. It's a bit harder get the frontend/backend to use it. The scope of this PR will be limited to switching our backend over to something that uses those types.

**Currently, we have the custom `Router` built, but our backend doesn't use it yet. One of the reasons it's not set up is because I'm too lazy to actually write the schemas for our APIs**

## Approach
Idea was inspired by [Gary Bernhardt](https://www.executeprogram.com/blog/porting-to-typescript-solved-our-api-woes
), but @jeffzh4ng and I couldn't figure out his implementation for the life of us. Instead we substituted a system of our own design. Maybe ours is better? We'll see about that when things are fully in place.

Alternatives? Here's one: https://hackernoon.com/express-fp-an-express-wrapper-for-type-safe-request-handlers-f8c411cc4a7b

> I might've jumped on making my own solution just for the sake of flexing and putting it on a resume

We use `io-ts` instead of Joi, which will eventually enable clean and easy (de)serialization. This seems to be a staple of anyone who tries to make shared types a thing.

## Testing
**TODO**

#### Open Questions and Pre-Merge TODOs
- [ ] Use github checklists. When solved, check the box and explain the answer.

## Learning
_Describe the research stage_

Lots of extremely arcane Typescript went into this. The official Typescript docs were the main source of information there.

_Links to blog posts, patterns, libraries or addons used to solve this problem_

#### Blog Posts
- [How to Pull Request](https://github.com/flexyford/pull-request) Github Repo with Learning focused Pull Request Template.

### CODE REVIEW

**Review by jeffzh4ng — CHANGES_REQUESTED** [review]
You're a Typescript wizard Justin 🧙‍♂️🔮

File: common/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  readonly response?: ResponseSchema
+}
+
+export type MethodName = 'all' | 'get' | 'post' | 'put' | 'delete' | 'patch' | 'options' | 'head'
+export type RootPath = '/'
+export type ChildPath = `/${string}`
+
+type Methods = Readonly<Partial<Record<MethodName, MethodSchema>>>
+type Children = Readonly<Partial<Record<Exclude<ChildPath, RootPath>, EndpointSchema>>>
+export interface EndpointSchema extends Methods {
```
[the remark this answers is not in the document]
Comment (xujustinj) [reply common/src/types.ts:22]: I think API makes sense for the root-level object, but what about something like **/clubs**? It makes more sense (to me, at least) to call that an endpoint

Maybe I can create a typealias `APISchema = EndpointSchema` so we get the best of both names 🤔

[the remark this answers is not in the document]
Comment (xujustinj) [reply common/src/types.ts:22]: Oops, that's leftover junk from my experimentation when I tried to flatten the object tree. I wanted "get" and "/clubs" to both be properties of the API schema, so we didn't have to put them under children/methods. It didn't work out; even though we're smart enough to distinguish whether there's a slash at the front of the property, TS couldn't

File: common/src/express.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+type TypedHandlerOf<
+  Method extends MethodSchema,
+  MethodType extends TypesOf<Method> = TypesOf<Method>
+> = RequestHandler<
+  MethodType extends { request: { parameters: {} } } ? MethodType['request']['parameters'] : {},
+  MethodType extends { response: { body: {} } } ? MethodType['response']['body'] : {},
+  MethodType extends { request: { body: {} } } ? MethodType['request']['body'] : {},
+  MethodType extends { request: { query: {} } } ? MethodType['request']['query'] : {}
+>
```
[the remark this answers is not in the document]
Comment (xujustinj) [reply common/src/express.ts:21]: The cost of that is we increase the number of exports

[the remark this answers is not in the document]
Comment (jeffzh4ng) [reply common/src/express.ts:21]: I think readability trumps number of exports for our web-framework code like this

File: common/src/demo.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    get: {
+      response: {
+        body: t.literal('Water water water, loo loo loo'),
+      },
+    },
+  },
+  children: {
+    '/clubs': clubsAPI,
+  },
+}
```
[the remark this answers is not in the document]
Comment (xujustinj) [reply common/src/demo.ts:34]: So we're just moving `.methods` into `.root.methods`? This is adding more nesting on top of what is already too much nesting

File: common/src/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (xujustinj) [reply common/src/types.ts]: > Wait..are `EndpointSchema` and `Children`'s definition mutually recursive?

Yep! The `/clubs` endpoint is also a fully-fledged Endpoint schema. `/clubs/:id` would be yet another layer in, but it also has an Endpoint schema. Altogether, it forms a tree, with our base API as the root.

That's why it's challenging to come up with a good name for things. I wouldn't really call the root API an "endpoint", but it quacks and walks like one. `/clubs/:id` sounds much more like an endpoint to me.

Since I never formally learned web dev, I don't really know what the proper terminology is. Are `/` and `/clubs` and `/club/:id` "endpoints" or "APIs"?

File: common/src/demo.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeffzh4ng) [reply common/src/demo.ts]: Yes it adds more nesting and makes the type defs more messy, but I'd argue the overall shape and readability of our `API` object will be cleaner and easier to grok. I'll let you make the final decision on this one though.

File: common/src/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeffzh4ng) [reply common/src/types.ts]: > Since I never formally learned web dev, I don't really know what the proper terminology is. Are / and /clubs and /club/:id "endpoints" or "APIs"?

My definition is the "API" is the structure, grouping, and schema of all "endpoints".

File: common/src/express/types.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  Get.Response.Body<Method>,
+  Get.Request.Body<Method>,
+  Get.Request.Query<Method>
+>
+
+export type TypedHandler<
+  Endpoint extends Schema.Endpoint,
+  Path extends Get.Path<Endpoint>,
+  MethodName extends Schema.MethodName = Schema.MethodName
+> = TypedHandlerForMethod<Get.ChildMethod<Endpoint, Path, MethodName>>
```
Comment (jeffzh4ng) [inline common/src/express/types.ts:16]: 👍

File: common/src/express/types.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  MethodName extends Schema.MethodName = Schema.MethodName
+> {
+  <Path extends Get.Path<Endpoint> & PathParams>(
+    path: Path,
+    ...handlers: Array<TypedHandler<Endpoint, Path, MethodName>>
+  ): ReturnType
+}
+
+export interface TypedRouter<Endpoint extends Schema.Endpoint> {
+  all: TypedMatcher<Endpoint, this, 'all'>
```
Comment (jeffzh4ng) [inline common/src/express/types.ts:30]: `Matcher` in the name `TypedMatcher` doesn't make any sense re: web development and API servers. The names of types that include `API`, `Endpoint`, `Router` etc make sense. `TypedMatcher` is too abstract, can we come up with a better name for this type?

**Review by jeffzh4ng — CHANGES_REQUESTED** [review]

File: common/src/express/router.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+  constructor(protected endpoint: Endpoint, options?: express.RouterOptions) {
+    this.router = express.Router(options)
+  }
+
+  public freeze(): this {
+    if (this.frozen) throw new FrozenRouterError()
+    this.frozen = true
+    return this
+  }
```
[the remark this answers is not in the document]
Comment (xujustinj) [reply common/src/express/router.ts:33]: I'm trying to model (in runtime) what we might eventually do in compile time. After you freeze a router, you shouldn't be able to modify it. A file might define a `clubRouter` and export `clubRouter.freeze()`, so that other files that import it can't modify it.

File: common/src/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (xujustinj) [reply common/src/types.ts]: Ok, then we can stick to "Endpoint".

Following your definition, maybe this makes sense:

```ts
type API = {
  root: Endpoint
}
```

[the remark this answers is not in the document]
Comment (jeffzh4ng) [reply common/src/types.ts]: Yeah I like that!

File: common/src/demo.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeffzh4ng) [reply common/src/demo.ts]: Yeah sounds good.

File: common/src/api-types.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    for (const key in endpoint.methods) {
+      // Typescript isn't smart enough to figure this out on its own
+      methodNames.add(key as MethodName<Endpoint>)
+    }
+    return methodNames
+  }
+
+  export type Path<Endpoint extends Schema.Endpoint> = Schema.Path &
+    (Endpoint extends { methods: Schema.Methods }
+      ? keyof Endpoint['children'] | Schema.RootPath
```
[the remark this answers is not in the document]
Comment (xujustinj) [reply common/src/api-types.ts:63]: 👀 nice catch, that's supposed to be an `&`

[the remark this answers is not in the document]
Comment (xujustinj) [reply common/src/api-types.ts:63]: So this is the weird thing about `Endpoint`. The `.methods` field corresponds to the methods available at the `/` endpoint. The `.children` field corresponds to the endpoints for the children. It's messy, but the `/` endpoint is fundamentally different from all the others in the sense that it can't have any further children.

`Schema.RootPath` is `/`, so this just means "if the endpoint has a `.methods`, then it supports the `/` path as well as all of its children"

File: common/src/express/router.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+  public all<Path extends Get.Path<Endpoint> & PathParams>(
```
[the remark this answers is not in the document]
Comment (xujustinj) [reply common/src/express/router.ts:35]: So `Get` is the (possibly misnamed, open to suggestions) namespace for the utility that gets things from endpoints.

`Get.Path<Endpoint>` is the type that represents all paths you can take from `Endpoint`. It doesn't have anything to do with the HTTP verb `GET`.

File: common/src/api-types.ts [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    readonly request?: Request
+    readonly response?: Response
+  }
+
+  export type MethodName = 'all' | 'get' | 'post' | 'put' | 'delete' | 'patch' | 'options' | 'head'
+  export type Methods = Readonly<Partial<Record<MethodName, Method>>>
+
+  export type Path = `/${string}`
+  export type RootPath = '/'
+  export type Children = Readonly<Partial<Record<Exclude<Path, RootPath>, Endpoint>>>
```
[the remark this answers is not in the document]
Comment (xujustinj) [reply common/src/api-types.ts:36]: I'm not sure actually, it's to signify that not every `Path` is guaranteed to be in `.children`. I'm not sure if this happens by default in `Record`, I didn't check

[the remark this answers is not in the document]
Comment (xujustinj) [reply common/src/api-types.ts:36]: Did some looking into it; the code works without the `Partial` here. However, if you remove `Partial` from the definition of the similar `Methods` type, it won't work. If your type has a finite number of values (like `MethodName`), then Typescript expects every value to be a property in the corresponding record. Removing the `Partial` in the definition of `Children` works because there are infinitely many possible `Path`s.

tl;dr: if you want a record type where all the keys are optional, `Partial<Record<K, V>>` _always_ works. Depending on `K`, you can sometimes get away with just doing `Record<K, V>`. To be safe, stick to always wrapping a `Partial` around your type, so that way your type definition works regardless of `K`.

File: common/src/api-types.ts [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    export type Body<Method extends Schema.Method> = TypeOf<
+      Field<Schema.Request, Response<Method>, 'body'>
+    >
+  }
+}
+
+function compiler<Base extends object>() {
+  return function <T extends Base>(obj: T): DeepReadonly<T> {
+    // No point calling Object.freeze, just let Typescript enforce the immutability
+    return obj as DeepReadonly<T>
```
Comment (jeffzh4ng) [inline common/src/api-types.ts:130]: 👍

**Review by jeffzh4ng — DISMISSED** [review]
